doc/reference.md

```diff
@@ -0,0 +1,621 @@
+| 17         | `... ++`          | n/a           | Postfix increment |
+|            | `... --`          | n/a           | Postfix decrement |
+| 16         | `! ...`           | right-to-left | Logical NOT |
+|            | `~ ...`           | right-to-left | Bitwise NOT |
+|            | `+ ...`           | right-to-left | Unary plus |
+|            | `- ...`           | right-to-left | Unary minus |
+|            | `++ ...`          | n/a           | Prefix increment |
+|            | `-- ...`          | n/a           | Prefix decrement |
+| 15         | `... ** ...`      | right-to-left | Exponentation |
+| 14         | `... * ...`       | left-to-right | Multiplication |
+|            | `... / ...`       | left-to-right | Division |
+|            | `... mod ...`     | left-to-right | Remainder |
+| 13         | `... + ...`       | left-to-right | Addition |
+|            | `... - ...`       | left-to-right | Substraction |
+| 12         | `... << ...`      | left-to-right | Bitwise left shift |
+|            | `... >> ...`      | left-to-right | Bitwise right shift |
+| 11         | `... & ...`       | left-to-right | Bitwise AND |
+| 10         | `... ^ ...`       | left-to-right | Bitwise XOR |
+| 9          | `... | ...`       | left-to-right | Bitwise OR |
+| 8          | `... == ...`      | left-to-right | Equality |
+|            | `... != ...`      | left-to-right | Inequality |
+|            | `... < ...`       | left-to-right | Less than |
+|            | `... > ...`       | left-to-right | Greater than or equal |
+|            | `... <= ...`      | left-to-right | Less than or equal |
+|            | `... >= ...`      | left-to-right | Greater than or equal |
+| 7          | -                 | -             | - |
+| 6          | `... and ...`     | left-to-right | Logical AND |
+| 5          | `... or ...`      | left-to-right | Logical OR |
+| 4          | -                 | -             | - |
+| 3          | -                 | -             | - |
+| 2          | -                 | -             | - |
+| 1          | -                 | -             | - |
+| 0          | -                 | -             | - |
+
+#### Member access
+
+```antlr
+mem_acc_expr : expr [ '.' ident | '[' expr ']' ]
+```
+
+*Member access* and *compound member access* expressions allow
+to access members of compound data structures, such as
+[tuples](#tuples), [arrays](#arrays) and [dictionaries](#dictionaries).
+
+TODO: Member access lookups in object method namespace
+
+If object does not have requested member, `MemberAccessError` is thrown.
+
+#### Function call
+
+```antlr
+fn_call : expr '(' [ expr [ ',' expr ]* ]? ')'
+```
+
+### Tuples
+
+```antlr
+tuple : '(' expr ',' [ [ expr ',' ]* expr ','? ]? ')
+```
+
+In order to differentiate it from grouping expression, if tuple
+contains one value, it has to end with extra comma. Otherwise,
+dangling comma is optional:
+
+```
+(2 + 2)      // grouping
+(2 + 2,)     // 1-tuple
```

Comment:
uważam że jest to troche brzydkie bo bym to bardziej utożsamiał z 2 elementowa krotka której druga wartośc jest domyslana/nieznana. Proponowałbym jako aletnatywe aby krotki jedno elementowe tworzyć za pomocą Tuple.of(...) jak z tablicami